File: linked_list/0092_reverse_linked_list_ii/0092_reverse_linked_list_ii.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def reverseBetween(self, head, m, n):
        diff, dummy, cur = n - m +1, ListNode(-1), head
        dummy.next = head

        last_unswapped = dummy
        while cur and m > 1:
            cur, last_unswapped, m = cur.next, cur, m-1
        
        prev, first_swapped = last_unswapped, cur
        while cur and diff > 0:
            cur.next, prev, cur, diff = prev, cur, cur.next, diff - 1

        last_unswapped.next, first_swapped.next = prev, cur

        return dummy.next


        

File: linked_list/0092_reverse_linked_list_ii/0092_reverse_linked_list_ii.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseBetween(head *ListNode, m int, n int) *ListNode {
	if head == nil || m >= n {
		return head
	}

	newHead := &ListNode{Val: 0, Next: head}
	pre := newHead
	for count := 0; pre.Next != nil && count < m-1; count++ {
		pre = pre.Next
	}
	if pre.Next == nil {
		return head
	}
	cur := pre.Next
	for i := 0; i < n-m; i++ {
		tmp := pre.Next
		pre.Next = cur.Next
		cur.Next = cur.Next.Next
		pre.Next.Next = tmp
	}

	return newHead.Next
}

// 1,2,3,4,5
// pre = 1, cur = 2
// loop 1:
// tmp = 2
// 1.next = 3
// 2.next = 2.next.next =4
// 1.next.next = 2

// 1,3,2,4,5
// pre=1, cur = 2
// loop2:
// tmp = 1.next = 3
// 1.next = 2.next = 4
// 2.next = 2.next.next =5
// 1.next.next = 3

// 1,4,3,2,5

